src/core/src/op/topk.cpp

```diff
@@ -104,59 +105,55 @@ bool evaluate_topk(const HostTensorPtr& arg,
     return rc;
 }
 
-template <element::Type_t K_ET>
-size_t get_k_from_hosttensor(const HostTensorPtr& arg) {
-    using T = typename element_type_traits<K_ET>::value_type;
-    auto p = arg->get_data_ptr<T>();
-    size_t k = p[0];
-    return k;
-}
-
-#define CASE_GET_K(a, ...)                                          \
-    case element::Type_t::a: {                                      \
-        OV_OP_SCOPE(OV_PP_CAT3(topk_get_k, _, a));                  \
-        k = get_k_from_hosttensor<element::Type_t::a>(__VA_ARGS__); \
-    } break
+template <element::Type_t INPUT_ET>
+bool no_bounds_overlap(const pair<HostTensorPtr, HostTensorPtr>& bounds,
+                       const size_t axis,
+                       const size_t k,
+                       const bool is_max_mode) {
+    using T = typename element_type_traits<INPUT_ET>::value_type;
 
-size_t read_k_from_host_tensor(const HostTensorPtr& arg_k) {
-    size_t k = 0;
-    switch (arg_k->get_element_type()) {
-        CASE_GET_K(i8, arg_k);
-        CASE_GET_K(i16, arg_k);
-        CASE_GET_K(i32, arg_k);
-        CASE_GET_K(i64, arg_k);
-        CASE_GET_K(u8, arg_k);
-        CASE_GET_K(u16, arg_k);
-        CASE_GET_K(u32, arg_k);
-        CASE_GET_K(u64, arg_k);
-    default:
-        // other types are not supported and would have thrown in ctor
-        ngraph_error("read_k_from_host_tensor: type is not integral\n");
-        break;
-    }
-    return k;
+    const auto k_max = shape_size(bounds.second->get_shape());
+    auto lowers = bounds.first->get_data_ptr<T>();
+    auto uppers = std::vector<T>(k_max);
+    auto indices = std::vector<int64_t>(k_max);
+
+    ngraph::runtime::reference::topk<T, int64_t>(bounds.second->get_data_ptr<T>(),
+                                                 indices.data(),
+                                                 uppers.data(),
+                                                 bounds.second->get_shape(),
+                                                 Shape({k_max}),
+                                                 axis,
+                                                 k_max,
+                                                 is_max_mode,
+                                                 ov::op::TopKSortType::SORT_VALUES);
+    auto prev_bound = is_max_mode ? std::numeric_limits<T>::max() : std::numeric_limits<T>::lowest();
+    size_t iteration = 0;
+
+    return std::find_if_not(uppers.cbegin(), uppers.cend(), [&](const T ub) {
+               const auto idx = indices[iteration];
+               auto no_overlap = (is_max_mode && ov::cmp::le(ub, prev_bound)) ||
+                                 (!is_max_mode && ov::cmp::le(prev_bound, lowers[idx]));
```

Comment:
As the dimension bounds are inclusive, the case with equal bounds is still an overlap.
Also there is a note in the [TopK-3](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/sort/TopK_3.md) specification:
"If there are several elements with the same value then their output order is not determined."
```suggestion
               auto no_overlap = (is_max_mode && ov::cmp::lt(ub, prev_bound)) ||
                                 (!is_max_mode && ov::cmp::lt(prev_bound, lowers[idx]));
```

Reply:
No, the lower then will not give correct results, there can be several dimension with same size (static size) and there will be no overlap ( will be sorted as reference implementation do)